Searchable Dropdown Multiselect

tuva-odegard opened the pull request:
## Beskrivelse

Fra [ønskelista](https://www.figma.com/design/uSg1Lz02to7HEZNzqBqi3p/%C3%98nskeliste?node-id=2645-3812) har det kommet inn et forslag om Multiselect. Det er nå gjort!

## Motivasjon og kontekst
Vi har to likende elementer: 
- Searchable dropdown
- Kontovelger/account selector multi

Kontovelger multi var kun tilgjengelig hvis dataen var kontoer.
Searchable dropdown tilbydde ikke  å velge flere. Searchable dropdown er nå oppdatert til en egen komponent som tilhører pakken, `<SearchableDropdownMultiSelect>`. 

## Testing
Det er testet at det fungerer med tastaturnavigasjon. Skjermleser-funksjonalitet er litt usikkert da den er litt ustabil på min maskin. 
Burde testes på små skjermer og med darkmode.

pethel commented on packages/ffe-searchable-dropdown-react/less/ffe-multiselect-searchable-dropdown.less:
```diff
@@ -0,0 +1,203 @@
+        border-radius: var(--ffe-g-border-radius);
+        transition: all var(--ffe-transition-duration) var(--ffe-ease);
+
+        &:has(.ffe-searchable-dropdown-multiselect--input-container .ffe-searchable-dropdown-multiselect--input-field:focus) {
```
Har du kjørt lint. Denne var veldig spesifik. Tror det blir noen klasser før mye

tuva-odegard replied:
Lint gir en warning om denne ja, vi kan gjerne se på sammen hvordan det går an å gjøre annerledes

pethel commented on packages/ffe-searchable-dropdown-react/less/ffe-multiselect-searchable-dropdown.less:
```diff
@@ -0,0 +1,203 @@
+        transition: all var(--ffe-transition-duration) var(--ffe-ease);
+
+        &:has(.ffe-searchable-dropdown-multiselect--input-container .ffe-searchable-dropdown-multiselect--input-field:focus) {
+            border-color: var(--ffe-v-input-bg-color);
```
Er det trygt att bruke v her @antidecaf. V er privat men på tvers av pakker?

antidecaf replied:
Det ligger en dependency til ffe-form i pakken, så den vil nok teknisk sett funke. Men det er i utgangspunktet meningen at `-v-`-variabler defineres innad i hver pakke og ikke gjenbrukes på tvers av pakker.

pethel commented on packages/ffe-searchable-dropdown-react/less/ffe-multiselect-searchable-dropdown.less:
```diff
@@ -0,0 +1,203 @@
+.ffe-searchable-dropdown-multiselect {
+    @check-icon-width: 24px;
+
+    &--container {
```
Vad med att droppa container og bare kalle containern `..ffe-searchable-dropdown-multiselect` ?

pethel commented on packages/ffe-searchable-dropdown-react/less/ffe-multiselect-searchable-dropdown.less:
```diff
@@ -0,0 +1,203 @@
+.ffe-searchable-dropdown-multiselect {
+    @check-icon-width: 24px;
```
css custom properties > less
```suggestion
    --check-icon-width: 24px;
```

pethel commented on packages/ffe-searchable-dropdown-react/less/ffe-multiselect-searchable-dropdown.less:
```diff
@@ -0,0 +1,203 @@
+        border-radius: var(--ffe-g-border-radius);
+        transition: all var(--ffe-transition-duration) var(--ffe-ease);
+
+        &:has(.ffe-searchable-dropdown-multiselect--input-container .ffe-searchable-dropdown-multiselect--input-field:focus) {
```
```suggestion
        &:has(.ffe-searchable-dropdown-multiselect--input-field:focus-visible) {
```

Focus visible slik att vi ikke får focus stilar på native når man klicker

tuva-odegard replied:
Har oppdatert til focus-visible. Det var et forslag om å fjerne .ffe-searchable-dropdown-multiselect--input-container, men det blir ikke riktig.

pethel commented on packages/ffe-searchable-dropdown-react/less/ffe-multiselect-searchable-dropdown.less:
```diff
@@ -0,0 +1,203 @@
+.ffe-searchable-dropdown-multiselect {
```
generelt må vi skilja på block og modifiers. --input-container og --input-field høres før meg ut som BLOCK og ikke modifiers
https://getbem.com/naming/

pethel commented on packages/ffe-searchable-dropdown-react/less/ffe-multiselect-searchable-dropdown.less:
```diff
@@ -0,0 +1,203 @@
+        }
+    }
+
+    &--input-container {
```
```suggestion
    &__input-container {
```

pethel commented on packages/ffe-searchable-dropdown-react/less/ffe-multiselect-searchable-dropdown.less:
```diff
@@ -0,0 +1,203 @@
+        font-size: var(--ffe-fontsize-form-input);
+    }
+
+    &--input-field {
```
```suggestion
    &__input-field {
```

pethel commented on packages/ffe-searchable-dropdown-react/src/MultiSelect/a11y.ts:
```diff
@@ -0,0 +1,170 @@
+import { useEffect, useRef } from 'react';
```
Synes ikke vi burde kopiera hele denne? Vad er forskjellen så kan jag prova hjelpe

pethel commented on packages/ffe-searchable-dropdown-react/less/ffe-multiselect-searchable-dropdown.less:
```diff
@@ -0,0 +1,203 @@
+            color: var(--ffe-g-primary-color);
+        }
+
+        &:focus {
```
```suggestion
        &:focus-visible {
```

pethel commented on packages/ffe-searchable-dropdown-react/less/ffe-multiselect-searchable-dropdown.less:
```diff
@@ -0,0 +1,203 @@
+        transition: border var(--ffe-transition-duration) var(--ffe-ease), border-left 0s;
+        color: var(--ffe-g-primary-color);
+
+        > .ffe-icons {
```
```suggestion
     .ffe-icons {
```

pethel commented on packages/ffe-searchable-dropdown-react/less/ffe-multiselect-searchable-dropdown.less:
```diff
@@ -0,0 +1,203 @@
+        }
+
+
+        &--flip {
```
Her (`--flip`) var det riktig og bruke modifier og ikke block :)

pethel commented on packages/ffe-searchable-dropdown-react/less/ffe-multiselect-searchable-dropdown.less:
```diff
@@ -0,0 +1,203 @@
+            border: 2px solid var(--ffe-g-primary-color);
+        }
+
+        &-icon {
```
```suggestion
        &__icon {
```

???

pethel commented on packages/ffe-searchable-dropdown-react/less/ffe-multiselect-searchable-dropdown.less:
```diff
@@ -0,0 +1,203 @@
+    &__list-item-body {
+        display: grid;
+        grid-column-gap: var(--ffe-spacing-sm);
+        grid-template-columns: 1fr @check-icon-width;
```
```suggestion
        grid-template-columns: 1fr var(--check-icon-width);
```

pethel commented on packages/ffe-searchable-dropdown-react/less/ffe-multiselect-searchable-dropdown.less:
```diff
@@ -0,0 +1,203 @@
+        }
+
+        @media (hover: hover) and (pointer: fine) {
+            &:hover, &:focus, &:active {
```
her vill focus og active kun slå in hvis man har muspeker. Dette active vill alltså ikke funke på mobil.

tuva-odegard replied:
Det går kanskje greit? Ikke sikker på hvordan det skal se ut på mobil

pethel commented on packages/ffe-searchable-dropdown-react/less/ffe-multiselect-searchable-dropdown.less:
```diff
@@ -0,0 +1,203 @@
+            }
+        }
+
+        &:focus, &:active {
```
```suggestion
        &:focus-visible, &:active {
```

pethel commented on packages/ffe-searchable-dropdown-react/less/ffe-multiselect-searchable-dropdown.less:
```diff
@@ -0,0 +1,203 @@
+            }
+        }
+
+        &-content {
```
```suggestion
        &__content {
```
???

pethel commented on packages/ffe-searchable-dropdown-react/less/ffe-multiselect-searchable-dropdown.less:
```diff
@@ -0,0 +1,203 @@
+.ffe-searchable-dropdown-multiselect {
```
Burde ikke vldig mye stilar kunna gjenbrukas mellom denne og searchable dropdown?

tuva-odegard replied:
Jeg prøvde.... Det var litt for mye ulikt dessverre

pethel commented on packages/ffe-searchable-dropdown-react/src/MultiSelect/Results.tsx:
```diff
@@ -0,0 +1,82 @@
+        isHighlighted: boolean;
+        isSelected: boolean;
+    }>;
+    selectedItem?: Item[] | Item | null;
```
selectedItems?: Item[]; ?

pethel commented on packages/ffe-searchable-dropdown-react/src/MultiSelect/ListItemContainer.tsx:
```diff
@@ -0,0 +1,60 @@
+import React, { useId, ForwardedRef } from 'react';
```
Her burde man kunna gjenbruka den vi har allrede? Alltså ListItemContainer? isSelected er forskjellen?

pethel commented on packages/ffe-searchable-dropdown-react/src/MultiSelect/Results.tsx:
```diff
@@ -0,0 +1,82 @@
+import { Locale } from '../types';
+import isEqual from 'lodash.isequal';
+
+interface ResultProps<Item extends Record<string, any>> {
```
PÅ den her hade jag også gjerne sett noe gjenbruk

pethel commented on packages/ffe-searchable-dropdown-react/src/MultiSelect/reducer.ts:
```diff
@@ -0,0 +1,142 @@
+    noMatchDropdownList?: Item[] | undefined;
+};
+
+export const createReducer =
```
Den her burde man også kunen gjenbruka en god del av eller er dom flesta olika her sammelignet med orginlaet?.

pethel commented on packages/ffe-searchable-dropdown-react/src/a11y.ts:
```diff
@@ -145,3 +145,24 @@ export const useSetAllyMessageItemSelection = ({
+    return dropdownAttributes.map(attr => highlightedValue[attr]).join(' ');
+};
+
+export const getArrowAllyMessage = <Item extends Record<string, any>>(
```
det her burde hete "setArrowAllyMessage"?

tuva-odegard committed:
refactor(ffe-account-selector-react): endret prop fra listelementBody til optionBody

BREAKING CHANGE: endret navn fra listElementBody til optionBody. Endring i.fbm searchabledropdownmultiselect

tuva-odegard committed:
feat(ffe-searchable-dropdown-react): Legger til multiselect.

BREAKING CHANGE: Endrer fra listElementBody til optionBody på eksisterende searchabledropdown

pethel commented on packages/ffe-searchable-dropdown-react/src/MultiSelect/MultiselectOptionBody.tsx:
```diff
@@ -0,0 +1,49 @@
+                )}
+            </div>
+            <Icon
+                fileUrl="data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIGhlaWdodD0iMjQiIHZpZXdCb3g9IjAgLTk2MCA5NjAgOTYwIiB3aWR0aD0iMjQiPgogICAgPHBhdGgKICAgICAgICBkPSJtMzgyLTM2Mi4xMzEgMzM0LjY5Ni0zMzQuNjk1UTczMC4zNy03MTAuNSA3NDguNzYxLTcxMC41dDMyLjA2NSAxMy42NzRRNzk0LjUtNjgzLjE1MiA3OTQuNS02NjQuMzhxMCAxOC43NzEtMTMuNjc0IDMyLjQ0NUw0MTQuMDY1LTI2NC40MTNRNDAwLjM5MS0yNTAuNzM5 ...
```
Kan vi gjøra typ 

const icon400SmNoe = "" ... slik att vi i framtiden vet vilken ikon som er brukt

pethel commented on packages/ffe-searchable-dropdown-react/src/MultiSelect/SearchableDropdownMultiSelect.stories.tsx:
```diff
@@ -0,0 +1,241 @@
+
+type Story = StoryObj<typeof SearchableDropdownMultiSelect>;
+
+export const Standard: Story = {
```
Nu kallar vi denne Standard. Jag har inga problem med det selv men vad tenker dere den burde hete?

pethel commented on packages/ffe-searchable-dropdown-react/src/MultiSelect/SearchableDropdownMultiSelect.stories.tsx:
```diff
@@ -0,0 +1,241 @@
+                    labelId={labelledById}
+                    inputId={id}
+                >
+                    <SearchableDropdownMultiSelect id="id" {...args} />
```
```suggestion
                    <SearchableDropdownMultiSelect {...args} />
```

pethel replied:
eller <SearchableDropdownMultiSelect id={id} {...args} />

pethel commented on packages/ffe-searchable-dropdown-react/src/MultiSelect/SearchableDropdownMultiSelect.stories.tsx:
```diff
@@ -0,0 +1,241 @@
+
+        return (
+            <InputGroup label="Velg frukt" labelId={labelledById} inputId={id}>
+                <SearchableDropdownMultiSelect
```
<SearchableDropdownMultiSelect
                    id={id}
                    labelledById={labelledById}
           
                    selectedItems={selectedOptions}
                    onChange={handleDropdownChange}
                       {...args}
                />
                
                Slik att man skriver over med det man sender in

tuva-odegard replied:
Men onChange er definert i args også, da mister man jo poenget med storyen? 
![image](https://github.com/user-attachments/assets/12e5ea06-9f06-4d5e-9cdf-35bf615d8fbb)

pethel replied:
ja ok ... .jag tenker at lite av poenget også er att kunna endre ting i stories men jag er ikke noen expert på storybook. Her fratar man iaf muligheterna til endring.

pethel commented on packages/ffe-searchable-dropdown-react/src/MultiSelect/SearchableDropdownMultiSelect.stories.tsx:
```diff
@@ -0,0 +1,241 @@
+        dropdownAttributes: ['displayName', 'color', 'amount'],
+    },
+    render: function Render({ id, labelledById, ...args }) {
+        const [selectedOptions, setSelectedOptions] = useState([
```
```
 const [selectedOptions, setSelectedOptions] = useState<typeof fruits>([
            fruits[0],
        ]);
```
eller noe sånt

tuva-odegard replied:
Hmm, da går det ikke å sende den inn til SearchableDropdownMultiSelect

pethel replied:
Vi ble enig på PM

pethel commented on packages/ffe-searchable-dropdown-react/src/Option.tsx:
```diff
@@ -36,17 +38,22 @@ function ListItemContainerWithForwardRef<Item extends Record<string, any>>(
             aria-selected={isSelected}
             ref={ref}
-            onClick={onClick}
+            onClick={e => {
```
onClick={onClick} eller så fjerne fordi du har den på rest her nedan. `e.preventDefault()` skall ikek trengas her vell eftersom det er en div bare man clicker på.

pethel commented on packages/ffe-searchable-dropdown-react/src/SingleSelect/SearchableDropdown.tsx:
```diff
@@ -366,70 +349,48 @@ function SearchableDropdownWithForwardRef<Item extends Record<string, any>>(
+                    />
+                )}
+                {postListElement && (
+                    <div className="ffe-searchable-dropdown__list--post-list-element">
```
Denne burde helst ligge utanfør lsiten eftersom det ikke er ett option. Skjønner dette er gammel kode. Hvis det har konsekvenser så trenger du ikke flytta den ut men går det smertfritt kan du gjerne legge den utanfør.

tuva-odegard replied:
Godt observert. men ja det ble ikke smertefritt

pethel commented on packages/ffe-searchable-dropdown-react/src/SingleSelect/reducer.ts:
```diff
@@ -146,7 +146,7 @@ export const createReducer =
                 if (shouldEmptySelectedItem) {
                     onChange?.(null);
-                    selectedItem = null;
+                    selectedItem = undefined;
```
Kan helst vare null lsik det var før meg. undefined er ting. Det er inget man assigner til ting.

tuva-odegard replied:
![image](https://github.com/user-attachments/assets/a4b6ffba-cb75-4241-8770-834303177bf8)

antidecaf commented:
Et aldri så lite UX-spørsmål:

Jeg velger noe fra nedtrekkslisten og skriver deretter fritekst. Når jeg så trykker backspace slettes ikke ett og ett tegn fra teksten slik man kanskje kan forvente, men alt jeg har skrevet i tillegg til siste chip som er lagt inn. Er dette forventet oppførsel?

![ezgif-6-d19c6212a8](https://github.com/user-attachments/assets/90c84e04-44eb-4a1f-8835-acd777a44648)

martingaard commented:
> Et aldri så lite UX-spørsmål:
> 
> Jeg velger noe fra nedtrekkslisten og skriver deretter fritekst. Når jeg så trykker backspace slettes ikke ett og ett tegn fra teksten slik man kanskje kan forvente, men alt jeg har skrevet i tillegg til siste chip som er lagt inn. Er dette forventet oppførsel?
> 
> ![ezgif-6-d19c6212a8](https://private-user-images.githubusercontent.com/463847/379734781-90c84e04-44eb-4a1f-8835-acd777a44648.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3Njc2NzIsIm5iZiI6MTcyOTc2NzM3MiwicGF0aCI6Ii80NjM4NDcvMzc5NzM0NzgxLTkwYzg0ZTA0LTQ0ZWItNGExZi04ODM1LWFjZDc3N2E0NDY0OC5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyNFQxMDU2MTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01OGRmYjRkZDg1ZjdmNGVhY2U5ZjA5N2M0NGE4YTBhYTc4YmEzYzk2YjQyNTdiYjliM2EyN2E1ODVmZjQ0OTNmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.is4MIkOfnTQTPDrI05tSqbFprRlidpDKE7TKY6Y8KxY) [ ![ezgif-6-d19c6212a8](https://private-user-images.githubusercontent.com/463847/379734781-90c84e04-44eb-4a1f-8835-acd777a44648.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3Njc2NzIsIm5iZiI6MTcyOTc2NzM3MiwicGF0aCI6Ii80NjM4NDcvMzc5NzM0NzgxLTkwYzg0ZTA0LTQ0ZWItNGExZi04ODM1LWFjZDc3N2E0NDY0OC5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyNFQxMDU2MTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01OGRmYjRkZDg1ZjdmNGVhY2U5ZjA5N2M0NGE4YTBhYTc4YmEzYzk2YjQyNTdiYjliM2EyN2E1ODVmZjQ0OTNmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.is4MIkOfnTQTPDrI05tSqbFprRlidpDKE7TKY6Y8KxY) ](https://private-user-images.githubusercontent.com/463847/379734781-90c84e04-44eb-4a1f-8835-acd777a44648.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3Njc2NzIsIm5iZiI6MTcyOTc2NzM3MiwicGF0aCI6Ii80NjM4NDcvMzc5NzM0NzgxLTkwYzg0ZTA0LTQ0ZWItNGExZi04ODM1LWFjZDc3N2E0NDY0OC5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyNFQxMDU2MTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01OGRmYjRkZDg1ZjdmNGVhY2U5ZjA5N2M0NGE4YTBhYTc4YmEzYzk2YjQyNTdiYjliM2EyN2E1ODVmZjQ0OTNmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.is4MIkOfnTQTPDrI05tSqbFprRlidpDKE7TKY6Y8KxY) [ ](https://private-user-images.githubusercontent.com/463847/379734781-90c84e04-44eb-4a1f-8835-acd777a44648.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3Njc2NzIsIm5iZiI6MTcyOTc2NzM3MiwicGF0aCI6Ii80NjM4NDcvMzc5NzM0NzgxLTkwYzg0ZTA0LTQ0ZWItNGExZi04ODM1LWFjZDc3N2E0NDY0OC5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyNFQxMDU2MTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01OGRmYjRkZDg1ZjdmNGVhY2U5ZjA5N2M0NGE4YTBhYTc4YmEzYzk2YjQyNTdiYjliM2EyN2E1ODVmZjQ0OTNmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.is4MIkOfnTQTPDrI05tSqbFprRlidpDKE7TKY6Y8KxY)

Nei, det er uheldig. Burde være mulig å trykke backspace

tuva-odegard commented:
> Et aldri så lite UX-spørsmål:
> 
> Jeg velger noe fra nedtrekkslisten og skriver deretter fritekst. Når jeg så trykker backspace slettes ikke ett og ett tegn fra teksten slik man kanskje kan forvente, men alt jeg har skrevet i tillegg til siste chip som er lagt inn. Er dette forventet oppførsel?
> 

Ikke forventet, bra funnet! 👍  Fikset nå.

tuva-odegard commented on packages/ffe-searchable-dropdown-react/src/MultiSelect/reducer.ts:
```diff
@@ -0,0 +1,187 @@
+                        highlightedIndex:
+                            action.payload?.highlightedIndex ?? -1,
+                    };
+                }
```
Dette er workaround for å sikre at man ikke har chip valgt samtidig som man kan gå nedover i lista. Håper ikke det brekker noe annet.

pethel replied:
Det er akkurat slika ting det er veldig fint att ha tester på :). Derfor jag er slitsom med det. Brekker man noe når man endrer. Det er greit att veta.
